### PR TITLE
ViT and CvT Models for Alzheimer Classification (ADNI Dataset) - Jack Fuller - 47446096

### PR DESCRIPTION
**Pull request by Fuller81**
# Overview

This pull request contains implementations of both the ViT architecture proposed in the study "An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale", and the CvT model proposed in the study "CvT: Introducing Convolutions to Vision Transformers". Both the ViT and CvT implementations were designed for the Alzheimer's Disease Neuroimaging Initiative dataset (ADNI for short). The dataset contains MRI scans of various subjects who have Alzheimer's (labelled as AD) and those who are normal controls (labelled as NC). The final Vit and CvT models had a final accuracy of 73.50% and 71.46% on the test dataset respectively. More information regarding the models, usage information, and results of this project can be found in the README file.

#  File Structure

- “modules.py": Contains the implementation of the ViT and CvT models. Each model consists of several components for each portion of the algorithm
- “dataset.py": Contains a variety of methods to perform initial data preprocessing and load train/val/test data loaders
- “train.py": Contains the training loop used for each of the models. The function prints progress updates after each epoch, plots train/val accuracy and loss, and additionally will save the model with either the latest validation accuracy or validation accuracy over 80%
- “predict.py": Contains a function that shows an example of using a saved model, and displays the accuracy, f1 score and confusion matrix of the predictions made.
- “README.MD”: Contains documentation for the project including information regarding the algorithms used, final results for the project, dependencies, data processing techniques and example usage methods and results.

### CODE REVIEW

**Comment by shaivikaaaa** [comment]
@gayanku

**Comment by aniketgupta17** [comment]
Observational Feedback 

File Organizing: Well-organized files.

Pull Request :
Correctly generated to Topic Recognition Branch .
The Description talks about the File Structure Clearly . 

Problem Solving:

The algorithm solves the problem appropriately but the result images must be included not just plots  . 
Accuracy in testing dataset: 73.50% and 71.46% 
Model and functions:

It correctly uses PyTorch to construct the models and functions.
Code design: Good.

Code comment and docstring:

Insufficient Code Commenting . 
Good function docstrings
Difficulty: Hard.

Additional Comments:

Good attempt at trying two versions of the model with clear results comparison.
Meaningful commits.
Good ReadMe design. However, It would be better if a conclusion and future improvements  appeared in the report.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. </td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. ViT (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Jack Fuller

**Comment by shakes76** [comment]
Well done
